datacite/index.js

```diff
@@ -0,0 +1,216 @@
+  if (isNaN(bufferSize)) { bufferSize = 1000; }
+  if (isNaN(throttle)) { throttle = 100; }
+  if (isNaN(ttl)) { ttl = 3600 * 24 * 7; }
+
+  if (!cache) {
+    const err = new Error('failed to connect to mongodb, cache not available for Datacite');
+    err.status = 500;
+    return err;
+  }
+
+  report.set('datacite', 'queries', 0);
+  report.set('datacite', 'query-fails', 0);
+  report.set('datacite', 'cache-fails', 0);
+
+  const process = bufferedProcess(this, {
+    packetSize,
+    bufferSize,
+
+    /**
+     * Filter ECs that should be enriched
+     * @param {Object} ec
+     * @returns {Boolean|Promise} true if the EC should be enriched, false otherwise
+     */
+    filter: ec => {
+      if (!ec.doi) { return false; }
+      if (!cacheEnabled) { return true; }
+
+      return findInCache(ec.doi).then(cachedDoc => {
+        if (cachedDoc) {
+          enrichEc(ec, cachedDoc);
+          return false;
+        }
+        return true;
+      });
+    },
+
+    onPacket: co.wrap(onPacket)
+  });
+
+  return new Promise(function (resolve, reject) {
+    // Verify cache indices and time-to-live before starting
+    cache.checkIndexes(ttl, function (err) {
+      if (err) {
+        logger.error(`Datacite: failed to verify indexes : ${err}`);
+        return reject(new Error('failed to verify indexes for the cache of Datacite'));
+      }
+
+      resolve(process);
+    });
+  });
+
+  /**
+     * Process a packet of ECs
+     * @param {Array<Object>} ecs
+     * @param {Map<String, Set<String>>} groups
+     */
+  function* onPacket({ ecs }) {
+    if (ecs.length === 0) { return; }
+
+    const dois = ecs.map(([ec, done]) => ec.doi);
+
+    const maxAttempts = 5;
+    let tries = 0;
+    let docs;
+
+    while (!docs) {
+      if (++tries > maxAttempts) {
+        const err = new Error(`Failed to query Datacite ${maxAttempts} times in a row`);
+        return Promise.reject(err);
+      }
+
+      try {
+        docs = yield query(dois);
+      } catch (e) {
+        logger.error(`Datacite: ${e.message}`);
+      }
+
+      yield wait(throttle);
+    }
+
+    const doiResults = new Map();
+    docs.forEach(doc => doiResults.set(doc.id, doc));
+
+    for (const [ec, done] of ecs) {
+      const doc = doiResults.get(ec.doi);
+
+      try {
+        // If we can't find a result for a given ID, we cache an empty document
+        yield cacheResult(ec.doi, doc || {});
+      } catch (e) {
+        report.inc('datacite', 'cache-fails');
+      }
+
+      if (doc) {
+        enrichEc(ec, doc);
+      }
+
+      done();
+    }
+  }
+
+  /**
+   * Enrich an EC using the result of a query
+   * @param {Object} ec the EC to be enriched
+   * @param {Object} result the document used to enrich the EC
+   */
+  function enrichEc(ec, result) {
+    const titleData = result.attributes.titles.find((res) => {
+      return res && res.title;
+    });
+
+    if (titleData) {
+      ec['publication_title'] = titleData.title;
+    }
```

Comment:
An exception may be thrown if the response is not as expected.

```suggestion
    const titles = result && result.attributes && result.attributes.titles;
    if (!Array.isArray(titles)) { return; }
    const titleData = titles.find((item) => (item && item.title));

    if (titleData) {
      ec['publication_title'] = titleData.title;
    }
```